airbyte-workers/src/main/java/io/airbyte/workers/helper/FailureHelper.java

```diff
@@ -120,11 +166,29 @@ private static Metadata jobAndAttemptMetadata(final Long jobId, final Integer at
         .withAdditionalProperty(ATTEMPT_NUMBER_METADATA_KEY, attemptNumber);
   }
 
+  private static Metadata traceMessageMetadata(final Long jobId, final Integer attemptNumber) {
+    return new Metadata()
+        .withAdditionalProperty(JOB_ID_METADATA_KEY, jobId)
+        .withAdditionalProperty(ATTEMPT_NUMBER_METADATA_KEY, attemptNumber)
+        .withAdditionalProperty(TRACE_MESSAGE_METADATA_KEY, true);
+  }
+
   /**
-   * Orders failures by timestamp, so that earlier failures come first in the list.
+   * Orders failures by putting errors from trace messages first, and then orders by timestamp, so
+   * that earlier failures come first.
    */
-  private static List<FailureReason> orderedFailures(final Set<FailureReason> failures) {
-    return failures.stream().sorted(Comparator.comparing(FailureReason::getTimestamp)).collect(Collectors.toList());
+  public static List<FailureReason> orderedFailures(final Set<FailureReason> failures) {
+    final Comparator<FailureReason> compareByIsTrace = Comparator.comparing(f -> {
+      final Object metadata = f.getMetadata();
+      if (metadata != null) {
+        return f.getMetadata().getAdditionalProperties().containsKey(TRACE_MESSAGE_METADATA_KEY) ? 0 : 1;
+      } else {
+        return 1;
+      }
+    });
+    final Comparator<FailureReason> compareByTimestamp = Comparator.comparing(FailureReason::getTimestamp);
+    final Comparator<FailureReason> compareByTraceAndTimestamp = compareByIsTrace.thenComparing(compareByTimestamp);
+    return failures.stream().sorted(compareByTraceAndTimestamp).collect(Collectors.toList());
```

Comment:
Nit: Since java 17 (or 18) we can use `.toList()` instead of `.collect(Collectors.toList())`.